### PR TITLE
Kernel - Fix subfolder property when using mount manager

### LINKED ISSUE #355: Mount Manager option show folder as Mounted Volume type
Feature request can skip this form. Bug report **must** complete it. `Check List` must be 100% match or it will be automatically closed without further discussion. Please remove this line.
# Environment

Windows version: Win 10
Processor architecture: x64
Dokany version: 1.0.0 RC4
Library type (Dokany/FUSE): Dokany
# Check List
- [ x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
- [x ] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
- [x ] I can always reproduce the issue with the provided description below.
- [x ] I have updated Dokany to the latest version (https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
# Description

When using Mount Manager and check Properties of a Folder you do NOT have Size, Size on disk, Contains props.
# Logs

Screenshot attached 
mountmanager

### PR DESCRIPTION
**Pull request by ATRiiX**
Fixes #355 . 
Compared with opening the properties of windows local drive's sub-directories, I found the IOCTL_MOUNTDEV_QUERY_DEVICE_NAME return STATUS_INVALID_PARAMETER, while the dokan mirror drive's sub-dir return SUCCESS.

### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the existing documentation
- [x] My changes generate no new warnings
- [ ] I have updated the change log (Add/Change/Fix)
- [x] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- add IsVolumeOpen check and return STATUS_INVALID_PARAMETER in the MOUNTDEVCONTROLTYPE.
-

### CODE REVIEW

**Comment by Liryna** [comment]
@ATRiiX Hi @ATRiiX,

Thanks for the PR!

Now that I read a little more the code, I believe the DiskDeviceControl should not include the MOUNTDEVCONTROLTYPE ioctl at all and have its own function that does the IsVolumeOpen check .
Like actually implement the switch case here with the two  functions and remove the IsVolumeOpen comment
https://github.com/dokan-dev/dokany/blob/3f7227cb7ba25c7caf78610daabd0e0376098745/sys/device.c#L739-L769

**Comment by ATRiiX** [comment]
Hi @Liryna , thanks for the reply.
I have update the pr with your suggestions. Now add IsVolumeOpen check when the code MOUNTDEVCONTROLTYPE, and test it can also fix the properties issue. Please take some time to review the pr.

File: sys/device.c [hunk 1]
```diff
@@ -730,38 +731,38 @@ NTSTATUS
...
-          RequestContext->IrpSp->Parameters.DeviceIoControl.IoControlCode);
-      // In case of IOCTL_STORAGE_BASE or IOCTL_DISK_BASE OR
-      // FILE_DEVICE_NETWORK_FILE_SYSTEM or MOUNTDEVCONTROLTYPE ioctl type, pass
-      // to DiskDeviceControl to avoid code duplication
-      // TODO: probably not the best way to pass down Irp...
-      if (baseCode == IOCTL_STORAGE_BASE || baseCode == IOCTL_DISK_BASE ||
-          baseCode == FILE_DEVICE_NETWORK_FILE_SYSTEM ||
-          baseCode == MOUNTDEVCONTROLTYPE) {
+      if (IsVolumeOpen(RequestContext->Vcb,
+                       RequestContext->IrpSp->FileObject)) {
```
Comment (Liryna) [inline sys/device.c:751]:
```
if (!IsVolumeOpen(...){
   status = STATUS_INVALID_PARAMETER;
   break;
}
status = DiskDeviceControlWithLock(....
```

**Review by Liryna — CHANGES_REQUESTED** [review]

**Comment by Liryna** [comment]
Thanks @ATRiiX for the quick change! I added some small comments and I believe we are good to merge it 😎

File: sys/device.c [hunk 2]
```diff
@@ -711,7 +711,8 @@ DiskDeviceControlWithLock(__in PREQUEST_CONTEXT RequestContext,
 // itself as opposed to a specific file.
 BOOLEAN
 IsVolumeOpen(__in PDokanVCB Vcb, __in PFILE_OBJECT FileObject) {
-  return FileObject != NULL && FileObject->FsContext == &Vcb->VolumeFileHeader;
+  return !(FileObject != NULL && FileObject->FsContext != NULL &&
+           FileObject->FsContext != &Vcb->VolumeFileHeader);
```
[the remark this answers is not in the document]
Comment (ATRiiX) [reply sys/device.c:715]: I added the log of FsContext on win11 and didn't see a case where it was equal to null. I think I've misremembered it. :)
So just keep the original judgment condition here.

File: sys/device.c (hunk 1 above)
Comment (ATRiiX) [reply sys/device.c:750]: Thanks, I have updated it.

**Comment by Liryna** [comment]
Thanks again for your contribution @ATRiiX 🏆 very appreciated!

**Pull request merged**